File: apps/system/api/domain/shared/address/address.go
package address

import "fmt"

type Address struct {
	id         ID
	zipCode    ZipCode
	country    Country
	prefecture Prefecture
	city       City
	street     *Street
	building   Building
}

func NewAddress(
	id ID,
	zipCode ZipCode,
	country Country,
	prefecture Prefecture,
	city City,
	street *Street,
	building Building,
) *Address {
	return &Address{
		id:         id,
		zipCode:    zipCode,
		country:    country,
		prefecture: prefecture,
		city:       city,
		street:     street,
		building:   building,
	}
}

func (a *Address) ToString() string {
	return fmt.Sprintf("%s %s %s %s %s %s", a.zipCode.ToString(), a.country.ToString(), a.prefecture.ToString(), a.city.ToString(), a.street.ToString(), a.building.ToString())
}

func (a *Address) ID() ID {
	return a.id
}

func (a *Address) ZipCode() ZipCode {
	return a.zipCode
}

func (a *Address) Country() Country {
	return a.country
}

func (a *Address) Prefecture() Prefecture {
	return a.prefecture
}

func (a *Address) City() City {
	return a.city
}

func (a *Address) Street() *Street {
	return a.street
}

func (a *Address) HasStreet() bool {
	return a.street != nil
}

func (a *Address) HasNotStreet() bool {
	return !a.HasStreet()
}

func (a *Address) Building() Building {
	return a.building
}

File: apps/system/client/src/domain/shared/address/address.ts
import { Entity } from "~/domain/shared"
import type {
  AddressBuilding,
  AddressCity,
  AddressCountry,
  AddressPrefecture,
  AddressStreet,
  AddressZipCode
} from "~/domain/shared/address"

interface AddressProps {
  zipCode: AddressZipCode
  country: AddressCountry
  prefecture: AddressPrefecture
  city: AddressCity
  street: AddressStreet | null
  building: AddressBuilding
}

export class Address extends Entity<AddressProps> {
  static create(v: AddressProps): Address {
    return new Address(v)
  }

  get zipCode(): AddressZipCode {
    return this.value.zipCode
  }

  get country(): AddressCountry {
    return this.value.country
  }

  get prefecture(): AddressPrefecture {
    return this.value.prefecture
  }

  get city(): AddressCity {
    return this.value.city
  }

  get street(): AddressStreet | null {
    return this.value.street
  }

  get building(): AddressBuilding {
    return this.value.building
  }
}
